lightning/src/ln/wire.rs

```diff
@@ -502,12 +525,63 @@ mod tests {
 		let mut reader = io::Cursor::new(buffer);
-		let decoded_msg = read(&mut reader).unwrap();
+		let decoded_msg = read(&mut reader, &IgnoringCustomMessageHandler{}).unwrap();
 		match decoded_msg {
 			Message::ChannelAnnouncement(msgs::ChannelAnnouncement { contents: msgs::UnsignedChannelAnnouncement { features, ..}, ..}) => {
 				assert!(!features.requires_unknown_bits());
 			},
 			_ => panic!("Expected node announcement, found message type: {}", decoded_msg.type_id())
 		}
 	}
+
+	#[derive(Eq, PartialEq, Debug)]
+	struct TestCustomMessage {}
+
+	const CUSTOM_MESSAGE_TYPE : u16 = 9000;
+
+	impl Type for TestCustomMessage {
+		fn type_id(&self) -> u16 {
+			CUSTOM_MESSAGE_TYPE
+		}
+	}
+
+	impl Writeable for TestCustomMessage {
+		fn write<W: Writer>(&self, _: &mut W) -> Result<(), io::Error> {
+			Ok(())
+		}
+	}
+
+	struct TestCustomMessageReader {}
+	impl core::ops::Deref for TestCustomMessageReader {
+		type Target = TestCustomMessageReader;
+		fn deref(&self) -> &Self { self }
+	}
+	impl CustomMessageReader for TestCustomMessageReader {
+		type CustomMessage = TestCustomMessage;
+		fn read<R: io::Read>(
+			&self,
+			message_type: u16,
+			_: &mut R
+		) -> Result<Option<Self::CustomMessage>, msgs::DecodeError> {
+			if message_type == CUSTOM_MESSAGE_TYPE {
+				return Ok(Some(TestCustomMessage{}));
+			}
+
+			Ok(None)
```

Comment:
Good point added. I wanted to do a more direct test using `TestCustomMessageHandler` directly but it complains about recursive references, so I'd need to drop the Deref or something, let me know if you think it's enough like that.

Reply:
Testing through `wire::read` is fine, but curious as to why you need to implement `Deref`? Taking a double reference like `&&TestCustomMessageReader {}` should work.

That said, does `wire::read` need to take a reference to something that implements `Deref`? Could it take a simple reference instead? I think `PeerManager` would need to dereference its custom handler and take a reference of the result for passing to `wire::read`.

Reply:
Ah good point about the double reference. Updated and added a direct test (kept the one through `read` as it was there).

The reason for using `Deref` is because if I just use a simple reference it complains that it cannot be made into an object due to the function using a generic parameter. Here is the error:

```
error[E0038]: the trait `CustomMessageReader` cannot be made into an object
   --> lightning/src/ln/wire.rs:122:17
    |
122 |     custom_reader: &CustomMessageReader<CustomMessage = T>,
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `CustomMessageReader` cannot be made into an object
    |
    = help: consider moving `read` to another trait
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
   --> lightning/src/ln/wire.rs:28:5
    |
21  | pub trait CustomMessageReader {
    |           ------------------- this trait cannot be made into an object...
...
28  |     fn read<R: io::Read>(&self, message_type: u16, buffer: &mut R) -> Result<Option<Self::CustomMessage>, msgs::DecodeError>;
    |        ^^^^ ...because method `read` has generic type parameters
```

Happy to hear if there is another way to do it though or if I'm misunderstanding something.

Reply:
Ah, that's right, it's a trait... sorry, I was confused. You can change the parameter type from `&H` to `H` and at the call site pass `&*self.custom_message_handler`.

Reply:
> You can change the parameter type from &H to H and at the call site pass &*self.custom_message_handler

Done